### PR TITLE
Trainable L matrix implemented

### PR DESCRIPTION
**Pull request by jwlee-jake**
Added trainable L on both predict_graph.py and construct_solution.py on the latest version. I also added the tau parameter when constructing the solution.

### CODE REVIEW

**Comment by Nish-19** [comment]
The changes look good to me.

File: create_solution.sh [hunk 1]
```diff
@@ -0,0 +1,7 @@
+#!/bin/bash
+
+for n in {420..650..15};
+do 
+    # echo $(($n))
+    python3 construct_solution.py -f final_10_27_20_25_29_final_stretch_batch_64_epoch_50_embed_300 -L $n &> tau_$n.log
```
Comment (wmcnicho) [inline create_solution.sh]: I don't know if we need to check this in

File: construct_solution.py [hunk 2]
```diff
@@ -102,11 +137,18 @@ def main():
...
             col_pos = construct_order_lst.index(col_cons)
-            if col_pos >= row_pos:
-                solution_adj_matrix[i][j] = 1 # construct j depends on construct i
+            solution_adj_matrix[i][j] = l_matrix[row_pos][col_pos]
+
 
     solution_adj_matrix_arr = np.array(solution_adj_matrix).astype(int)
-    np.save(f'./submissions/adj_matrix_{options.file_name}.npy', solution_adj_matrix_arr)
+
+    directory = f"tau_{options.tau / 1000}"
```
Comment (wmcnicho) [inline construct_solution.py]: Let's change this to use float

File: predict_graph.py [hunk 3]
```diff
@@ -23,14 +23,29 @@
...
+
 class PermutationMatrix(nn.Module):
     def __init__(self, input_size, temperature, unroll, verbose=False):
         super().__init__()
         self.unroll, self.temperature = unroll, temperature
         self.matrix = nn.Parameter(torch.empty(input_size, input_size, device=device))
         nn.init.kaiming_uniform_(self.matrix, a=math.sqrt(input_size))
-        self.lower = nn.Parameter(torch.tril(torch.ones(input_size, input_size)), requires_grad=False)
+
+        # NOTE: Trainable L.
```
Comment (wmcnicho) [inline predict_graph.py:35]: Remove all dead code

File: predict_graph.py [hunk 4]
```diff
@@ -47,12 +62,30 @@ def forward(self, epoch):
...
+        #     torch.set_printoptions(threshold=10_000)
+        #     print("init_l:\n", init_l)
+        #     lower = torch.sigmoid(self.lower * init_l) * l_mask
+        #     init_mask = init_mask + 1
+        # else:
+        #     print("Lower matrix through sigmoid")
+        #     lower = torch.sigmoid(self.lower) * l_mask
+        lower = torch.sigmoid(self.lower) * l_mask
+        torch.set_printoptions(threshold=100_000)
+        print(">> self.lower:\n", self.lower)
```
Comment (wmcnicho) [inline predict_graph.py]: put these under debug flag

File: create_solution.sh (hunk 1 above)
Comment (jwlee-jake) [reply create_solution.sh]: I'll put this in .gitignore & write down on README.md to run the code as following
python3 construct_solution.py -f final_10_27_20_25_29_final_stretch_batch_64_epoch_50_embed_300 -L 0.45

File: construct_solution.py (hunk 2 above)
Comment (jwlee-jake) [reply construct_solution.py]: removed 1000.

File: predict_graph.py (hunk 4 above)
Comment (jwlee-jake) [reply predict_graph.py]: I put those under verbose flag.

File: predict_graph.py (hunk 3 above)
Comment (jwlee-jake) [reply predict_graph.py:35]: removed all dead codes except for the accuracy part.

**Comment by jwlee-jake** [comment]
@Nish-19 Hey can you take a look at the predict_graph.py for removing dead codes?
Especially, with TODO and accuracy part. 
I have cleaned up as much as possible.

**Comment by Nish-19** [comment]
I have removed the dead code. This version looks good to me. Please have a look.

File: README.md [hunk 5]
```diff
@@ -1,9 +1,40 @@
...
+```
+
+Here is an example how to create a submission file. We submitted L value of 0.45.
+```
+python3 construct_solution.py -f ./submissions/final/final_10_27_20_25_29_final_stretch_batch_64_epoch_50_embed_300 -L 0.45
+```
 
-```predict_graph.py``` - Our code
\ No newline at end of file
+Contact: ml4ed @ UMass Amherst
```
Comment (wmcnicho) [inline README.md:52]: Maybe we should put our umass emails?

File: README.md [hunk 6]
```diff
@@ -1,9 +1,40 @@
-# Neurips Challenge 2022
+# Causal Knowledge Tracing
+NeurIPS 2022 CausalEdu Competition - Task 3
+
+## Setups
+The requiring environment is as bellow:
```
Comment (wmcnicho) [inline README.md]: 👍 let's add a `requirements.txt` here as well

File: create_solution.sh (hunk 1 above)
Comment (wmcnicho) [inline create_solution.sh]: Should we remove this?

Comment (jwlee-jake) [reply create_solution.sh]: Yeah you can remove it.

File: construct_solution.py [hunk 7]
```diff
@@ -46,6 +47,7 @@ def main():
     parser.add_argument('-D', '--debug', action=argparse.BooleanOptionalAction, help='Controls whether intermediate files are created during submission construction')
     parser.add_argument('-T', '--temperature', type=int, default=2, help='temperature of learned model')
     parser.add_argument('-U', '--unroll', type=int, default=5, help='unroll length of learned model')
+    parser.add_argument('-L', '--tau', type=float, default=0.5, help='threshold for L matrix')
```
Comment (wmcnicho) [inline construct_solution.py]: nitpick: we could change this to `0.45` or whatever is the experimentally best default

File: README.md [hunk 8]
```diff
@@ -1,9 +1,40 @@
...
+- PyTorch 1.12.1 
+- Numpy 1.23.1
+- Wandb 0.13.4
 
 ## Repository Overview 
 
 ```/data/``` - Repository containing the dataset
 
-```aritra_start.py``` - Reference code written by Aritra for causal discovery 
+```/serialized_torch/``` - <TODO: Hunter, can you help me writing this?>
```
Comment (wmcnicho) [inline README.md]: 👍 Will do

File: construct_solution.py (hunk 7 above)
Comment (jwlee-jake) [reply construct_solution.py]: Yeah, that will work

**Comment by wmcnicho** [comment]
GOING LIVE!
